zcash_client_sqlite/src/wallet/db.rs

```diff
@@ -216,54 +223,68 @@ CREATE TABLE orchard_received_note_spends (
     UNIQUE (orchard_received_note_id, transaction_id)
 )";
 
-/// Stores the current UTXO set for the wallet, as well as any transparent outputs
-/// previously observed by the wallet.
+/// Stores the transparent outputs received by the wallet.
 ///
 /// Originally this table only stored the current UTXO set (as of latest refresh), and the
 /// table was cleared prior to loading in the latest UTXO set. We now upsert instead of
 /// insert into the database, meaning that spent outputs are left in the database. This
-/// makes it similar to the `*_received_notes` tables in that it can store history, but
-/// has several downsides:
-/// - The table has incomplete contents for recovered-from-seed wallets.
-/// - The table can have inconsistent contents for seeds loaded into multiple wallets
+/// makes it similar to the `*_received_notes` tables in that it can store history.
+/// Depending upon how transparent TXOs for the wallet are discovered, the following
+/// may be true:
+/// - The table may have incomplete contents for recovered-from-seed wallets.
+/// - The table may have inconsistent contents for seeds loaded into multiple wallets
 ///   simultaneously.
-/// - The wallet's transparent balance can be incorrect prior to "transaction enhancement"
+/// - The wallet's transparent balance may be incorrect prior to "transaction enhancement"
 ///   (downloading the full transaction containing the transparent output spend).
-pub(super) const TABLE_UTXOS: &str = r#"
-CREATE TABLE "utxos" (
+///
+/// ### Columns:
+/// - `id`: Primary key
+/// - `transaction_id`: Reference to the transaction in which this TXO was created
+/// - `output_index`: The output index of this TXO in the transaction referred to by `transaction_id`
+/// - `account_id`: The account that controls spend authority for this TXO
+/// - `address`: The address to which this TXO was sent
+/// - `script`: The full txout script
+/// - `value_zat`: The value of the TXO in zatoshis
+/// - `max_observed_unspent_height`: The maximum block height at which this TXO was observed
+///   to be a member of the UTXO set. This is intended to be used to determine when the TXO
+///   is no longer a part of the UTXO set, in the case that the transaction that spends it is
+///   not detected by the wallet.
```

Comment:
This is still unclear about off-by-one issues.
```suggestion
/// - `max_observed_unspent_height`: The maximum block height at which this TXO was either
///   observed to be a member of the UTXO set at the start of the block, or observed to be an
///   output of a transaction mined in the block. This is intended to be used to determine when
///   the TXO is no longer a part of the UTXO set, in the case that the transaction that spends it
///   is not detected by the wallet.
```
I'm still worried by the fact that we currently never seem to use the contents of this field, and so I don't know whether these semantics will be what we actually want.

Reply:
We need to be able to detect when one of our UTXOs is no longer part of the UTXO set; since this table is being updated, I think that now is a good time to start populating this field so that we can later add the functionality to identify when we need to go searching for a fully-transparent spend of one of our transparent UTXOs, which we can't otherwise detect.